python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/scenario_state.py

```diff
@@ -0,0 +1,347 @@
+from dagster._core.definitions.events import CoercibleToAssetKey
+from dagster._core.definitions.executor_definition import in_process_executor
+from dagster._core.definitions.internal_asset_graph import InternalAssetGraph
+from dagster._core.definitions.repository_definition.repository_definition import (
+    RepositoryDefinition,
+)
+from dagster._core.definitions.repository_definition.valid_definitions import (
+    SINGLETON_REPOSITORY_NAME,
+)
+from dagster._core.host_representation.external_data import external_repository_data_from_def
+from dagster._core.host_representation.origin import InProcessCodeLocationOrigin
+from dagster._core.instance import DagsterInstance
+from dagster._core.storage.tags import PARTITION_NAME_TAG
+from dagster._core.test_utils import (
+    InProcessTestWorkspaceLoadTarget,
+    create_test_daemon_workspace_context,
+)
+from dagster._core.types.loadable_target_origin import LoadableTargetOrigin
+from dagster._serdes.utils import create_snapshot_id
+from dagster._seven.compat.pendulum import pendulum_freeze_time
+from typing_extensions import Self
+
+from .base_scenario import run_request
+
+FAIL_TAG = "test/fail"
+
+
+def get_code_location_origin(
+    scenario_spec: "ScenarioSpec",
+    location_name="test_location",
+    repository_name=SINGLETON_REPOSITORY_NAME,
+) -> InProcessCodeLocationOrigin:
+    """Hacky method to allow us to point a code location at a module-scoped attribute, even though
+    the attribute is not defined until the scenario is run.
+    """
+    repository = create_repository_using_definitions_args(
+        name=repository_name,
+        assets=scenario_spec.assets,
+        executor=in_process_executor,
+        sensors=scenario_spec.sensors,
+    )
+
+    return _get_code_location_origin_from_repository(
+        cast(RepositoryDefinition, repository), location_name=location_name
+    )
+
+
+def _get_code_location_origin_from_repository(repository: RepositoryDefinition, location_name: str):
+    attribute_name = (
+        f"_asset_daemon_target_{create_snapshot_id(external_repository_data_from_def(repository))}"
+    )
+
+    if attribute_name not in globals():
+        globals()[attribute_name] = repository
+
+    return InProcessCodeLocationOrigin(
+        loadable_target_origin=LoadableTargetOrigin(
+            executable_path=sys.executable,
+            module_name=("dagster_tests.definitions_tests.auto_materialize_tests.scenario_state"),
+            working_directory=os.getcwd(),
+            attribute=attribute_name,
+        ),
+        location_name=location_name,
+    )
+
+
+class AssetSpecWithPartitionsDef(
+    namedtuple(
+        "AssetSpecWithPartitionsDef",
+        AssetSpec._fields + ("partitions_def",),
+        defaults=(None,) * (1 + len(AssetSpec._fields)),
+    )
+):
+    ...
+
+
+class MultiAssetSpec(NamedTuple):
+    specs: Sequence[AssetSpec]
+    partitions_def: Optional[PartitionsDefinition] = None
+    can_subset: bool = False
```

Comment:
These classes and the code around them are kind of bananas and we should use `AssetNode` instead with @smackesey lands it